src/source/src/txn_channel.rs

```diff
@@ -0,0 +1,159 @@
+    )
+}
+
+/// Semaphore-based permits to control the back-pressure.
+///
+/// The number of messages in the transaction channel is limited by these semaphores.
+#[derive(Debug)]
+pub struct Permits {
+    /// The permits for records in chunks.
+    records: Semaphore,
+}
+
+impl Permits {
+    /// Add permits back to the semaphores.
+    pub fn add_permits(&self, permit_value: PermitValue) {
+        self.records.add_permits(permit_value.0 as usize)
+    }
+}
+
+/// The sender of the transaction channel with permit-based back-pressure.
+#[derive(Debug, Clone)]
+pub struct Sender {
+    pub tx: mpsc::UnboundedSender<TxnMsgWithPermits>,
+    permits: Arc<Permits>,
+
+    /// The maximum permits required by a chunk. If there're too many rows in a chunk, we only
+    /// acquire these permits. [`BATCHED_PERMITS`] is subtracted to avoid deadlock with
+    /// batching.
+    max_chunk_permits: usize,
+}
+
+impl Sender {
+    /// Send a message, waiting until there are enough permits.
+    /// Used to send transaction data messages.
+    ///
+    /// Returns error if the receive half of the channel is closed, including the message passed.
+    pub async fn send(
+        &self,
+        txn_msg: TxnMsg,
+        notificator: oneshot::Sender<usize>,
+    ) -> Result<(), mpsc::error::SendError<TxnMsg>> {
+        // The semaphores should never be closed.
+        let permits = match &txn_msg {
+            TxnMsg::Data(_, c) => {
+                let card = c.cardinality().clamp(1, self.max_chunk_permits);
+                if card == self.max_chunk_permits {
+                    tracing::warn!(
+                        cardinality = c.cardinality(),
+                        "large chunk in transaction channel"
+                    )
+                }
+                self.permits
+                    .records
+                    .acquire_many(card as _)
+                    .await
+                    .unwrap()
+                    .forget();
+                Some(PermitValue(card as _))
+            }
+            TxnMsg::Begin(_) | TxnMsg::Rollback(_) | TxnMsg::End(_) => None,
+        };
+
+        self.tx
+            .send(TxnMsgWithPermits {
+                txn_msg,
+                notificator,
+                permit_value: permits,
+            })
+            .map_err(|e| mpsc::error::SendError(e.0.txn_msg))
+    }
+
+    /// Send a message without permit acquiring.
+    /// Used to send transaction control messages.
+    ///
+    /// Returns error if the receive half of the channel is closed, including the message passed.
+    pub fn send_raw(
+        &self,
+        txn_msg: TxnMsg,
+        notificator: oneshot::Sender<usize>,
+    ) -> Result<(), mpsc::error::SendError<TxnMsg>> {
+        self.tx
+            .send(TxnMsgWithPermits {
+                txn_msg,
+                notificator,
+                permit_value: None,
+            })
+            .map_err(|e| mpsc::error::SendError(e.0.txn_msg))
+    }
```

Comment:
Perhaps we can use `now_or_never` here to ensure there's no misuse.

```suggestion
    pub fn send_immediate(
        &self,
        txn_msg: TxnMsg,
        notificator: oneshot::Sender<usize>,
    ) -> Result<(), mpsc::error::SendError<TxnMsg>> {
        self.send(..).now_or_never().expect("cannot send immediately")
    }
```

Reply:
But this `send` is not a `async` function.

Reply:
```
pub fn send(&self, message: T) -> Result<(), SendError<T>>
```

Reply:
Yes. That's what `now_or_never` for.

https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never